src/utils/focus/focusable.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * CSS selector to query focusable elements.
+ *
+ * This does not eliminate the following elements which are not focusable:
+ *  - Custom elements with `tabindex` or `contenteditable` attribute
+ *  - Shadow hosts with `delegatesFocus: true`
+ */
+export const FOCUSABLE_SELECTOR = [
+  'input:not([type=hidden]):not([disabled])',
+  'button:not([disabled])',
+  'select:not([disabled])',
+  'textarea:not([disabled])',
+  '[contenteditable=""]',
+  '[contenteditable="true"]',
+  'a[href]',
+  '[tabindex]:not([disabled])',
+].join(', ')
+
+/**
+ * Determine if an element can be the target for `focusElement()`.
+ *
+ * This does not necessarily mean that this element will be the `activeElement`,
+ * as it might delegate focus into a shadow tree.
+ */
+export function isFocusTarget(element: Element): element is HTMLElement {
+  if (element.tagName.includes('-')) {
+    // custom elements without `delegatesFocus` are ignored
+    return delegatesFocus(element)
+  }
+  // elements without `delegatesFocus` behave normal even if they're a shadow host
+  return delegatesFocus(element) || element.matches(FOCUSABLE_SELECTOR)
+}
+
+export function isFocusable(element: Element): element is HTMLElement {
+  return (
+    !element.tagName.includes('-') &&
+    !delegatesFocus(element) &&
+    element.matches(FOCUSABLE_SELECTOR)
+  )
+}
+
+export function delegatesFocus(
+  element: Element,
+): element is HTMLElement & {shadowRoot: ShadowRoot & {delegatesFocus: true}} {
+  // `delegatesFocus` is missing in Jsdom
+  // see https://github.com/jsdom/jsdom/issues/3418
+  // We'll treat `undefined` as `true`
+  return (
+    !!element.shadowRoot &&
```

Comment:
wouldn't  `element.shadowRoot` be sufficient here? The `!!` shouldn't make a difference right?